File: windows/net/samples/MultiAudioCD/ProgressForm.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using PrimoSoftware.Burner;


namespace MultiAudioCD
{
	public class ProgressForm: System.Windows.Forms.Form
	{
		public ProgressForm()
		{
			InitializeComponent();
		}

		protected override void Dispose(bool disposing)
		{
			if(disposing)
				if(components != null)
					components.Dispose();

			base.Dispose(disposing);
		}

		public bool Stopped 
		{ 
			get 
			{ 
				return m_stopped; 
			} 
		}

        void UIThread(MethodInvoker code)
        {
            if (InvokeRequired)
            {
                Invoke(code);
                return;
            }

            code.Invoke();
        }

		public void UpdateProgress(ProgressInfo info, int rowIndex) 
		{
            MethodInvoker del = delegate
            {
                while (lvDevices.Items.Count <= rowIndex)
                {
                    ListViewItem lviNewEntry = new ListViewItem();
                    lviNewEntry.SubItems.Add(string.Empty);
                    lviNewEntry.SubItems.Add(string.Empty);
                    lviNewEntry.SubItems.Add(string.Empty);
                    lvDevices.Items.Add(lviNewEntry);
                }

                ListViewItem lvi = lvDevices.Items[rowIndex];

                lvi.SubItems[0].Text = info.DeviceTitle;
                lvi.SubItems[1].Text = info.Status;
                lvi.SubItems[2].Text = info.ProgressStr;
                lvi.SubItems[3].Text = info.WriteSpeed;
            };

            UIThread(del);
		}

		private void buttonStop_Click(object sender, System.EventArgs e)
		{
			buttonStop.Enabled = false;
			m_stopped = true;
		}

		private bool m_stopped = false;

		#region Windows Form Designer generated code
        private System.Windows.Forms.Button buttonStop;
        private ListView lvDevices;
        private ColumnHeader columnHeaderDevice;
        private ColumnHeader columnHeaderStatus;
        private ColumnHeader columnHeaderProgress;
        private ColumnHeader columnHeaderWriteSpeed;

		/// <summary>
		/// Required designer variable.
		/// </summary>
		private System.ComponentModel.Container components = null;

		/// <summary>
		/// Required method for Designer support - do not modify
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent()
		{
            this.buttonStop = new System.Windows.Forms.Button();
            this.lvDevices = new System.Windows.Forms.ListView();
            this.columnHeaderDevice = ((System.Windows.Forms.ColumnHeader)(new System.Windows.Forms.ColumnHeader()));
            this.columnHeaderStatus = ((System.Windows.Forms.ColumnHeader)(new System.Windows.Forms.ColumnHeader()));
            this.columnHeaderProgress = ((System.Windows.Forms.ColumnHeader)(new System.Windows.Forms.ColumnHeader()));
            this.columnHeaderWriteSpeed = ((System.Windows.Forms.ColumnHeader)(new System.Windows.Forms.ColumnHeader()));
            this.SuspendLayout();
            // 
            // buttonStop
            // 
            this.buttonStop.Location = new System.Drawing.Point(687, 169);
            this.buttonStop.Name = "buttonStop";
            this.buttonStop.Size = new System.Drawing.Size(71, 20);
            this.buttonStop.TabIndex = 1;
            this.buttonStop.Text = "Stop";
            this.buttonStop.Click += new System.EventHandler(this.buttonStop_Click);
            // 
            // lvDevices
            // 
            this.lvDevices.Columns.AddRange(new System.Windows.Forms.ColumnHeader[] {
            this.columnHeaderDevice,
            this.columnHeaderStatus,
            this.columnHeaderProgress,
            this.columnHeaderWriteSpeed});
            this.lvDevices.FullRowSelect = true;
            this.lvDevices.GridLines = true;
            this.lvDevices.Location = new System.Drawing.Point(7, 5);
            this.lvDevices.Name = "lvDevices";
            this.lvDevices.Size = new System.Drawing.Size(757, 148);
            this.lvDevices.TabIndex = 2;
            this.lvDevices.UseCompatibleStateImageBehavior = false;
            this.lvDevices.View = System.Windows.Forms.View.Details;
            // 
            // columnHeaderDevice
            // 
            this.columnHeaderDevice.Text = "Device";
            this.columnHeaderDevice.Width = 363;
            // 
            // columnHeaderStatus
            // 
            this.columnHeaderStatus.Text = "Status";
            this.columnHeaderStatus.Width = 218;
            // 
            // columnHeaderProgress
            // 
            this.columnHeaderProgress.Text = "Progress";
            this.columnHeaderProgress.Width = 77;
            // 
            // columnHeaderWriteSpeed
            // 
            this.columnHeaderWriteSpeed.Text = "Write Speed";
            this.columnHeaderWriteSpeed.Width = 88;
            // 
            // ProgressForm
            // 
            this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
            this.ClientSize = new System.Drawing.Size(770, 203);
            this.ControlBox = false;
            this.Controls.Add(this.lvDevices);
            this.Controls.Add(this.buttonStop);
            this.FormBorderStyle = System.Windows.Forms.FormBorderStyle.FixedDialog;
            this.MaximizeBox = false;
            this.MinimizeBox = false;
            this.Name = "ProgressForm";
            this.ShowInTaskbar = false;
            this.StartPosition = System.Windows.Forms.FormStartPosition.CenterScreen;
            this.Text = "Working ...";
            this.ResumeLayout(false);

		}
		#endregion
	}
}

File: windows/net/samples.vb/MultiDataDisc/ProgressForm.vb
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports System.Windows.Forms
Imports PrimoSoftware.Burner


Namespace MultiDataDisc
	Public Class ProgressForm
		Inherits System.Windows.Forms.Form

		Public Sub New()
			InitializeComponent()
		End Sub

		Protected Overrides Sub Dispose(ByVal disposing As Boolean)
			If disposing Then
				If components IsNot Nothing Then
					components.Dispose()
				End If
			End If

			MyBase.Dispose(disposing)
		End Sub

		Public ReadOnly Property Stopped() As Boolean
			Get
				Return m_stopped
			End Get
		End Property

		Private Sub UIThread(ByVal code As MethodInvoker)
			If InvokeRequired Then
				Invoke(code)
				Return
			End If

			code.Invoke()
		End Sub

		Public Sub UpdateProgress(ByVal info As ProgressInfo, ByVal rowIndex As Integer)
			Dim del As MethodInvoker = Function() AnonymousMethod1(rowIndex, info)

			UIThread(del)
		End Sub
		
		Private Function AnonymousMethod1(ByVal rowIndex As Integer, ByVal info As MultiDataDisc.ProgressInfo) As Object
			Do While lvDevices.Items.Count <= rowIndex
				Dim lviNewEntry As New ListViewItem()
				lviNewEntry.SubItems.Add(String.Empty)
				lviNewEntry.SubItems.Add(String.Empty)
				lviNewEntry.SubItems.Add(String.Empty)
				lvDevices.Items.Add(lviNewEntry)
			Loop
			Dim lvi As ListViewItem = lvDevices.Items(rowIndex)
			lvi.SubItems(0).Text = info.DeviceTitle
			lvi.SubItems(1).Text = info.Status
			lvi.SubItems(2).Text = info.ProgressStr
			lvi.SubItems(3).Text = info.WriteSpeed
			Return Nothing
		End Function

		Private Sub buttonStop_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles buttonStop.Click
			buttonStop.Enabled = False
			m_stopped = True
		End Sub

		Private m_stopped As Boolean = False

		#Region "Windows Form Designer generated code"
		Private WithEvents buttonStop As System.Windows.Forms.Button
		Private lvDevices As ListView
		Private columnHeaderDevice As ColumnHeader
		Private columnHeaderStatus As ColumnHeader
		Private columnHeaderProgress As ColumnHeader
		Private columnHeaderWriteSpeed As ColumnHeader

		''' <summary>
		''' Required designer variable.
		''' </summary>
		Private components As System.ComponentModel.Container = Nothing

		''' <summary>
		''' Required method for Designer support - do not modify
		''' the contents of this method with the code editor.
		''' </summary>
		Private Sub InitializeComponent()
			Me.buttonStop = New System.Windows.Forms.Button()
			Me.lvDevices = New System.Windows.Forms.ListView()
			Me.columnHeaderDevice = (CType(New System.Windows.Forms.ColumnHeader(), System.Windows.Forms.ColumnHeader))
			Me.columnHeaderStatus = (CType(New System.Windows.Forms.ColumnHeader(), System.Windows.Forms.ColumnHeader))
			Me.columnHeaderProgress = (CType(New System.Windows.Forms.ColumnHeader(), System.Windows.Forms.ColumnHeader))
			Me.columnHeaderWriteSpeed = (CType(New System.Windows.Forms.ColumnHeader(), System.Windows.Forms.ColumnHeader))
			Me.SuspendLayout()
			' 
			' buttonStop
			' 
			Me.buttonStop.Location = New System.Drawing.Point(687, 169)
			Me.buttonStop.Name = "buttonStop"
			Me.buttonStop.Size = New System.Drawing.Size(71, 20)
			Me.buttonStop.TabIndex = 1
			Me.buttonStop.Text = "Stop"
'			Me.buttonStop.Click += New System.EventHandler(Me.buttonStop_Click)
			' 
			' lvDevices
			' 
			Me.lvDevices.Columns.AddRange(New System.Windows.Forms.ColumnHeader() { Me.columnHeaderDevice, Me.columnHeaderStatus, Me.columnHeaderProgress, Me.columnHeaderWriteSpeed})
			Me.lvDevices.FullRowSelect = True
			Me.lvDevices.GridLines = True
			Me.lvDevices.Location = New System.Drawing.Point(7, 5)
			Me.lvDevices.Name = "lvDevices"
			Me.lvDevices.Size = New System.Drawing.Size(757, 148)
			Me.lvDevices.TabIndex = 2
			Me.lvDevices.UseCompatibleStateImageBehavior = False
			Me.lvDevices.View = System.Windows.Forms.View.Details
			' 
			' columnHeaderDevice
			' 
			Me.columnHeaderDevice.Text = "Device"
			Me.columnHeaderDevice.Width = 363
			' 
			' columnHeaderStatus
			' 
			Me.columnHeaderStatus.Text = "Status"
			Me.columnHeaderStatus.Width = 218
			' 
			' columnHeaderProgress
			' 
			Me.columnHeaderProgress.Text = "Progress"
			Me.columnHeaderProgress.Width = 77
			' 
			' columnHeaderWriteSpeed
			' 
			Me.columnHeaderWriteSpeed.Text = "Write Speed"
			Me.columnHeaderWriteSpeed.Width = 88
			' 
			' ProgressForm
			' 
			Me.AutoScaleBaseSize = New System.Drawing.Size(5, 13)
			Me.ClientSize = New System.Drawing.Size(770, 203)
			Me.ControlBox = False
			Me.Controls.Add(Me.lvDevices)
			Me.Controls.Add(Me.buttonStop)
			Me.FormBorderStyle = System.Windows.Forms.FormBorderStyle.FixedDialog
			Me.MaximizeBox = False
			Me.MinimizeBox = False
			Me.Name = "ProgressForm"
			Me.ShowInTaskbar = False
			Me.StartPosition = System.Windows.Forms.FormStartPosition.CenterScreen
			Me.Text = "Working ..."
			Me.ResumeLayout(False)

		End Sub
		#End Region
	End Class
End Namespace
